src/checkers/inference/solver/frontend/ConstraintSerializer.java

```diff
@@ -34,13 +35,15 @@
     private Serializer<S, T> realSerializer;
 
     @SuppressWarnings("unchecked")
-    public ConstraintSerializer(String backEndType, Lattice lattice) {
+    public ConstraintSerializer(BackEndType backEndType, Lattice lattice) {
         try {
-            if (backEndType.equals(Constants.LINGELING)) {
+            String backEndPath = backEndType.getFullyQualifiedName();
+            if (backEndType.getBackEndType().equals(BackEndTypeEnum.Lingeling)) {
```

Comment:
Serializer and Back end was decoupled. One of the reasons we made the change is the current design is easier to implement custom serialization logic. For example, in type system X, it needs its own custom serialization logic for combine constraint, what we need to do is making a subclass of ConstraintSerializer, and override the serialize method for CombineConstraint. However, if we let back end create its own serializer, we will need to also make a subclass of backend. Moreover, the current serializer design is more friendly to multiple serialization methods. If we have multiple serialization approaches for CombineConstraint of type system X, for example, maxSAT and logiql, we can put them all in XConstraintSerializer.